packages/@react-aria/autocomplete/src/useAutocomplete.ts

```diff
@@ -0,0 +1,146 @@
+import {InputHTMLAttributes, KeyboardEvent, ReactNode, useRef} from 'react';
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import {useLocalizedStringFormatter} from '@react-aria/i18n';
+import {useTextField} from '@react-aria/textfield';
+
+export interface AriaAutocompleteProps extends AutocompleteProps, DOMProps, InputDOMProps, AriaLabelingProps {
+  children: ReactNode
+}
+
+// TODO: all of this is menu specific but will need to eventually be agnostic to what collection element is inside
+// Update all instances of "menu" then
+export interface AriaAutocompleteOptions extends Omit<AriaAutocompleteProps, 'children'> {
+  /** The ref for the input element. */
+  inputRef: RefObject<HTMLInputElement | null>
+}
+export interface AutocompleteAria<T> {
+  /** Props for the label element. */
+  labelProps: DOMAttributes,
+  /** Props for the autocomplete input element. */
+  inputProps: InputHTMLAttributes<HTMLInputElement>,
+  /** Props for the menu, to be passed to [useMenu](useMenu.html). */
+  menuProps: AriaMenuOptions<T>,
+  /** Props for the autocomplete description element, if any. */
+  descriptionProps: DOMAttributes,
+  // TODO: fairly non-standard thing to return from a hook, discuss how best to share this with hook only users
+  // This is for the user to register a callback that upon recieving a keyboard event key returns the expected virtually focused node id
+  /** Register function that expects a callback function that returns the newlly virtually focused menu option when provided with the keyboard action that occurs in the input field. */
+  register: (callback: (e: KeyboardEvent) => string | null) => void
+}
+
+/**
+ * Provides the behavior and accessibility implementation for a autocomplete component.
+ * A autocomplete combines a text input with a menu, allowing users to filter a list of options to items matching a query.
+ * @param props - Props for the autocomplete.
+ * @param state - State for the autocomplete, as returned by `useAutocompleteState`.
+ */
+export function useAutocomplete<T>(props: AriaAutocompleteOptions, state: AutocompleteState): AutocompleteAria<T> {
+  let {
+    inputRef,
+    isReadOnly
+  } = props;
+
+  let menuId = useId();
+
+  // TODO: may need to move this into Autocomplete? Kinda odd to return this from the hook? Maybe the callback should be considered
+  // external to the hook, and the onus is on the user to pass in a onKeydown to this hook that updates state.focusedNodeId in response to a key
+  // stroke
+  let callbackRef = useRef<(e: KeyboardEvent) => string>(null);
+  let register = (callback) => {
+    callbackRef.current = callback;
+  };
+
+  // For textfield specific keydown operations
+  let onKeyDown = (e: BaseEvent<KeyboardEvent<any>>) => {
+    if (e.nativeEvent.isComposing) {
+      return;
+    }
+
+    // TODO: how best to trigger the focused element's action? Currently having the registered callback handle dispatching a
+    // keyboard event
+    switch (e.key) {
```

Comment:
At the moment the keyboard event dispatch handling happens in Menu directly, but it might be good to have those in a hook somewhere. Its a bit tough since we don't have access to the wrapped collection and collection ref at the top level, but potentially could just have the callbackRef expect to receive the collection and collection ref instead of just the id, open to opinions here